debug/clusterinfo/clusterinfo.go

```diff
@@ -0,0 +1,94 @@
+// Copyright © 2021 Alibaba Group Holding Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package clusterinfo
+
+import (
+	"context"
+	"sort"
+
+	"github.com/pkg/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	corev1client "k8s.io/client-go/kubernetes/typed/core/v1"
+)
+
+// GetPodsIP returns the pods IP without duplicates.
+func GetPodsIP(ctx context.Context, client corev1client.CoreV1Interface, namespace string) ([]string, error) {
+	ipList := []string{}
+	podList, err := client.Pods(namespace).List(ctx, metav1.ListOptions{})
+	if  err != nil {
+		return nil, errors.Wrapf(err, "failed to get pods ip")
+	}
+
+	for i, _ := range podList.Items {
+		ipList = append(ipList, podList.Items[i].Status.PodIP)
+	}
+
+	ipList = removeDuplicatesAndEmpty(ipList)
+
+	return ipList, nil
+}
+
+// GetNodesIP returns the nodes IP.
+func GetNodesIP(ctx context.Context, client corev1client.CoreV1Interface) ([]string, error) {
+	ipList := []string{}
+	nodeList, err := client.Nodes().List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get nodes ip")
+	}
+
+	for i, _ := range nodeList.Items {
+		ipList = append(ipList, nodeList.Items[i].Status.Addresses[0].Address)
+	}
+
+	return ipList, nil
+}
+
+// GetDNSServiceAll return the DNS Service domain、DNS service cluster IP、DNS service endpoints.
+func GetDNSServiceAll(ctx context.Context, client corev1client.CoreV1Interface) (string, string, []string, error) {
+	namespace := "kube-system"
+	serviceName := "kube-dns"
+	domain := "kube-dns.kube-system.svc"
+
+	service, err := client.Services(namespace).Get(ctx, serviceName, metav1.GetOptions{})
+	if err != nil {
+		return "", "", nil, errors.Wrapf(err, "failed to get DNS service")
+	}
+	serviceClusterIP := service.Spec.ClusterIP
+
+	endpoints, err := client.Endpoints(namespace).Get(ctx, serviceName, metav1.GetOptions{})
+	if err != nil {
+		return "", "", nil, errors.Wrapf(err, "failed to get DNS service endpoint")
+	}
+	endpointsIPs := []string{}
+	for _, address := range endpoints.Subsets[0].Addresses {
+		endpointsIPs = append(endpointsIPs, address.IP)
+	}
+
+	return domain, serviceClusterIP, endpointsIPs, nil
+}
+
+func removeDuplicatesAndEmpty(ss []string) []string {
```

Comment:
Please add a unit test for this function.

Reply:
Got it

Reply:
Thanks. From the time complexity, using map to remove repetition is O(n), and I will think about using it. But you say the results seems not true, can you give some examples？The result of your given example is true.

Reply:
I'm not sure you sorts the silce.... may the result need to be sorted ?! 

when add `sort.Strings` method to `removeRep`.   perfrom is bad...

```
func removeRep(a []string) []string {
	sort.Strings(a)
	if len(a) == 0 {
		return a
	}
	res := make([]string, 0, len(a))
	tmp := map[string]struct{}{}
	for _, v := range a {
		if _, ok := tmp[v]; !ok && len(v) != 0 {
			tmp[v] = struct{}{}
			res = append(res, v)
		}
	}
	return res
}

then test  Benchmark_removeRep

pkg: github.com/fanux/sealos/k8s
cpu: Intel(R) Core(TM) i7-4850HQ CPU @ 2.30GHz
Benchmark_removeRep
Benchmark_removeRep-8   	 2358082	       512.8 ns/op	     184 B/op	       2 allocs/op
PASS
ok  	github.com/fanux/sealos/k8s	1.752s

pkg: github.com/fanux/sealos/k8s
cpu: Intel(R) Core(TM) i7-4850HQ CPU @ 2.30GHz
Benchmark_remove01
Benchmark_remove01-8   	 2418132	       487.6 ns/op	     136 B/op	       4 allocs/op
PASS
ok  	github.com/fanux/sealos/k8s	1.708s
```

Reply:
The results do not need to be sorted. The sorted in begin was for removing repetition.

Would it be better to change `tmp := map[string]struct{}{}` to `tmp := map[string]bool`?

Reply:
> Would it be better to change tmp := map[string]struct{}{} to tmp := map[string]bool?

yes. here you go.